First implementation of REST API task

a1exanddrovich opened the pull request.

Commit:
First implementation of REST API task

NatalliaKhadunai commented on pom.xml:
```diff
@@ -0,0 +1,149 @@
+
+    <!-- https://mvnrepository.com/artifact/org.thymeleaf/thymeleaf-spring5 -->
+    <dependency>
+      <groupId>org.thymeleaf</groupId>
```
I believe we don't need `thymeleaf` dependency as it's usually used when an app has UI. We don't have UI so we don't need this dependency.
Please remove it if I'm correct.

a1exanddrovich replied:
You're absolutely right because we don't have UI in the project. Then we don't need that one

NatalliaKhadunai commented on pom.xml:
```diff
@@ -0,0 +1,149 @@
+
+  <build>
+    <finalName>CertificateSystem</finalName>
+    <pluginManagement><!-- lock down plugins versions to avoid using Maven defaults (may be moved to parent pom) -->
```
Why do we need to list all these plugins?

a1exanddrovich replied:
These plugins allow the reusing of common build logic across this project. For example, they are able to create a WAR file or compile unit tests. So we need all the plugins

AliakseiYermachonak replied:
can't you package war or compile without this plugin declaration?

a1exanddrovich replied:
As I learnt I can:)
We can do it the way I Implemented just using maven-war-plugin dependency

NatalliaKhadunai commented on src/main/java/com/epam/esm/config/DispatcherServletInitializer.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Override
+    protected Class<?>[] getRootConfigClasses() {
+        return null;
```
If you're using IntelliJ IDEA please pay attention to the warnings it provides. In this case a warning looks like this:
![image](https://user-images.githubusercontent.com/17936204/130677387-9619cc6b-f3b9-4379-9ce7-21e905ad1b18.png)

a1exanddrovich replied:
As I got that hint was actually produced by the SonarLint plugin. I haven't installed this one by now

NatalliaKhadunai commented on src/main/java/com/epam/esm/config/SpringConfig.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    @Bean
+    public SpringResourceTemplateResolver templateResolver() {
```
I don't think we need beans related to a view resolution as we don't have UI. You can remove beans `templateResolver`, `templateEngine`, `configureViewResolvers`

a1exanddrovich replied:
Fixed

NatalliaKhadunai commented on src/main/java/com/epam/esm/config/SpringConfig.java:
```diff
@@ -0,0 +1,132 @@
+        DriverManagerDataSource dataSource = new DriverManagerDataSource();
+
+        dataSource.setDriverClassName("com.mysql.cj.jdbc.Driver");
+        dataSource.setUrl("jdbc:mysql://localhost:3306/giftcertificatesdatabase?serverTimezone=UTC");
```
Please move all the properties to a property file

a1exanddrovich replied:
Props moved to a property file, fixed

NatalliaKhadunai commented on src/main/java/com/epam/esm/config/SpringConfig.java:
```diff
@@ -0,0 +1,132 @@
+
+    @Bean
+    public TagService tagService() {
+        return new TagService(tagDao(), giftCertificateTagDao());
```
I think you can make use of `@Component` and `@Service` annotations instead of creating service, dao beans here in the config

a1exanddrovich replied:
Fixed

NatalliaKhadunai commented on src/main/java/com/epam/esm/controller/TagsController.java:
```diff
@@ -0,0 +1,55 @@
+@RequestMapping("/tags")
+public class TagsController {
+
+    private final static String JSON = "application/json";
```
![image](https://user-images.githubusercontent.com/17936204/130677945-89789e4e-2822-4409-acdc-9d571608aa71.png)

a1exanddrovich replied:
Modifiers reordered

NatalliaKhadunai commented on src/main/java/com/epam/esm/controller/TagsController.java:
```diff
@@ -0,0 +1,55 @@
+    public ResponseEntity<List<Tag>> getTags() {
+        List<Tag> tags = service.getTags();
+
+        return tags.size() == 0 ? new ResponseEntity<>(HttpStatus.NO_CONTENT)
```
It would be better to return an HTTP status 200 OK and `{}` as a response body when all the resources are requested and none were found.

a1exanddrovich replied:
Done

NatalliaKhadunai commented on src/main/java/com/epam/esm/controller/TagsController.java:
```diff
@@ -0,0 +1,55 @@
+    public ResponseEntity<Tag> getTag(@PathVariable("id") long id) {
+        Optional<Tag> tag = service.getTag(id);
+
+        return tag.isEmpty() ? new ResponseEntity<>(HttpStatus.NO_CONTENT)
```
It's better to return 404 NOT FOUND status when a single resource is requested and it can't be found

a1exanddrovich replied:
Fixed

NatalliaKhadunai commented on src/main/java/com/epam/esm/service/TagService.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public void createTag(Tag tag) throws BadEntityException {
+        if (tag == null) {
```
I believe that `tag` is never going to be `null` here. Probably we can remove the check

a1exanddrovich replied:
That's true, I should make a check correctly

NatalliaKhadunai commented on src/main/java/com/epam/esm/controller/TagsController.java:
```diff
@@ -0,0 +1,55 @@
+    @PostMapping(produces = JSON)
+    public ResponseEntity<Tag> createTag(@RequestBody Tag tag) throws BadEntityException {
+        service.createTag(tag);
+        return new ResponseEntity<>(HttpStatus.NO_CONTENT);
```
It's better to return 201 CREATED when a resource is created successfully. In a response body, we should return either a JSON representation of a created entity or its id.

a1exanddrovich replied:
Got, implemented

NatalliaKhadunai commented on src/main/java/com/epam/esm/controller/GiftCertificatesController.java:
```diff
@@ -0,0 +1,65 @@
+    @PatchMapping(value = "/{id}", produces = JSON)
+    public ResponseEntity<GiftCertificate> updateGiftCertificate(@PathVariable("id") long id, @RequestBody GiftCertificate giftCertificate) throws EntityNotExistException, BadEntityException {
+        service.updateGiftCertificate(id, giftCertificate);
+        return new ResponseEntity<>(giftCertificate, HttpStatus.OK);
```
You should better return an updated gift certificate here. Not the one that is being passed in a request body

a1exanddrovich replied:
Fixed

NatalliaKhadunai commented on src/main/java/com/epam/esm/controller/BusinessExceptionHandler.java:
```diff
@@ -0,0 +1,41 @@
+
+    @ExceptionHandler(BadEntityException.class)
+    public ResponseEntity<ExceptionResponse> handleBadEntityException(Locale locale) {
+        String errorMessage = messageSource.getMessage("exception.badEntity", new Object[]{}, locale);
```
Methods `handleEntityNotExistException` and `handleBadEntityException` are almost identical. Please try to avoid code duplication and move common code somewhere

a1exanddrovich replied:
Created a new private method instead od duplications

NatalliaKhadunai commented on src/main/java/com/epam/esm/entity/GiftCertificate.java:
```diff
@@ -0,0 +1,111 @@
+    private String creationDate;
+    private String lastUpdateDate;
+    @JsonProperty
+    private List<String> tags = new ArrayList<>();
```
We should have `List<Tag>` here

a1exanddrovich replied:
Fixed

NatalliaKhadunai commented on src/main/java/com/epam/esm/entity/GiftCertificate.java:
```diff
@@ -0,0 +1,111 @@
+    private String name;
+    private String description;
+    private BigDecimal price;
+    private int duration;
```
Don't forget to use `Duration` class

a1exanddrovich replied:
Yea, I remember about it

Commit:
Sql scripts added

NatalliaKhadunai commented on src/main/java/com/epam/esm/service/GiftCertificateService.java:
```diff
@@ -0,0 +1,116 @@
+@Component
+public class GiftCertificateService {
+
+    private final static String MINSK_TIME = "GMT+3:00";
```
It would be better to create a property to represent a timezone and store it in a property file

a1exanddrovich replied:
Moved

NatalliaKhadunai commented on src/main/java/com/epam/esm/service/GiftCertificateService.java:
```diff
@@ -0,0 +1,116 @@
+        }
+
+        giftCertificate.setLastUpdateDate(DATE_FORMAT.format(new Date()));
+        String updateQuery = constructor.constructUpdateQuery(id, giftCertificate);
```
It would be better to make calls to the `QueryConstructor` in dao's, not in services

a1exanddrovich replied:
Moved to the Dao layer

NatalliaKhadunai commented on src/main/java/com/epam/esm/service/GiftCertificateService.java:
```diff
@@ -0,0 +1,116 @@
+            if (!tagNames.contains(tag)) {
+                tagDao.create(new Tag(tag));
+            }
+        }
```
Lines 90-96 should be reimplemented to use Java 8 streams

a1exanddrovich replied:
Implemented

NatalliaKhadunai commented on src/main/java/com/epam/esm/dao/ConjugationDao.java:
```diff
@@ -0,0 +1,52 @@
+        template.update(DELETE_FROM_GIFT_CERTIFICATE_TAG, id);
+    }
+
+    public void deleteTagByIdFromGiftTag(long id) {
```
It would be better to rename the method to smth like `deleteByTagId`

a1exanddrovich replied:
Renamed

NatalliaKhadunai commented on src/main/java/com/epam/esm/dao/ConjugationDao.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.List;
+
+@Component
+public class ConjugationDao {
```
Please rename this class to `CertificateTagDao`. I have no idea what conjugation means...

a1exanddrovich replied:
Renamed

NatalliaKhadunai commented on src/main/java/com/epam/esm/utils/QueryConstructor.java:
```diff
@@ -0,0 +1,79 @@
+public class QueryConstructor {
+
+    public String constructQuery(String tagName, String giftCertificateName, String description, String sortByName, String sortByDate) {
+        String query = "SELECT gc.id, gc.name, gc.description, gc.price, gc.duration, gc.create_date, gc.last_update_date, gct.gift_certificate_id, t.name AS tagName FROM gift_certificate gc JOIN gift_certificate_tag gct ON gct.gift_certificate_id = gc.id JOIN tag t ON gct.tag_id = t.id";
```
We have the same query in the `GiftCertificateDao`. Please reuse it

a1exanddrovich replied:
Fixed

NatalliaKhadunai commented on src/main/java/com/epam/esm/utils/QueryConstructor.java:
```diff
@@ -0,0 +1,79 @@
+        }
+
+        //if (giftCertificate.getPrice() > 0) {
+        if (giftCertificate.getPrice().compareTo(BigDecimal.ZERO) >= 1) {
```
Please remove the comment if it's not needed

a1exanddrovich replied:
Removed

NatalliaKhadunai commented on src/main/java/com/epam/esm/utils/QueryConstructor.java:
```diff
@@ -0,0 +1,79 @@
+            query += "duration = " + giftCertificate.getDuration() + ", ";
+        }
+
+        query += "last_update_date = '" + giftCertificate.getLastUpdateDate() + "' ";
```
I believe we need to populate `last_update_date` with a new value whenever we're updating a gift certificate

a1exanddrovich replied:
The field is updated every time we call the method

NatalliaKhadunai reviewed (changes requested):
Some things missing:
* Validation

Commit:
Some bugfixes performed and new classes added(validators, dtos, etc.)

Commit:
tests added

Commit:
some fixes

Commit:
profiles added

NatalliaKhadunai commented on scripts/init-data.sql:
```diff
@@ -0,0 +1,101 @@
+create database giftcertificatedatabase;
```
Please remove the file if it's not used. init.sql is enough

a1exanddrovich replied:
Removed

NatalliaKhadunai commented on scripts/init.sql:
```diff
@@ -0,0 +1,30 @@
+(
+    id bigint auto_increment,
+    name varchar(50),
+    description varchar(50),
```
I believe that the field description should be bigger than 50 symbols.

a1exanddrovich replied:
To be honest  in real database I use it's actually has 100 symbols. Fixed

NatalliaKhadunai commented on scripts/init.sql:
```diff
@@ -0,0 +1,30 @@
+    price decimal(10, 2) default 0 null,
+    duration int,
+    create_date varchar(50),
+    last_update_date varchar(50),
```
Your DB should have types for representing date, time, datetime data. Please use suitable type, but no varchar

a1exanddrovich replied:
Now it does :)

NatalliaKhadunai commented on scripts/init.sql:
```diff
@@ -0,0 +1,30 @@
+create table tag
+(
+    id bigint auto_increment,
+    name varchar(50),
```
As we agreed this field should be unique in the DB

a1exanddrovich replied:
It's unique now

NatalliaKhadunai commented on src/main/java/com/epam/esm/dao/GiftCertificateDao.java:
```diff
@@ -0,0 +1,70 @@
+
+    @Override
+    public Optional<GiftCertificate> findById(long id) {
+        return template.query(SqlQueries.FIND_CERTIFICATE_BY_ID, mapper, new Object[]{id}).stream().findAny();
```
Sonarlint says that we can catch NullPointer here:
![image](https://user-images.githubusercontent.com/17936204/131492559-29f6cb5f-efd0-48f1-a3af-5ce46c2f1121.png)
And I'm not sure why do we need to apply these methods .stream().findAny() here?

a1exanddrovich replied:
We need the methods because it returns a List, while we need Optional to be returned

AliakseiYermachonak replied:
and why cannot we have null check?

a1exanddrovich replied:
Added Objects.requireNonNull method

NatalliaKhadunai commented on src/main/java/com/epam/esm/entity/ExceptionResponse.java:
```diff
@@ -0,0 +1,22 @@
+package com.epam.esm.entity;
```
This class should be located in some other package. It's not an entity

a1exanddrovich replied:
Moved it to exception package

NatalliaKhadunai commented on src/main/java/com/epam/esm/entity/GiftCertificate.java:
```diff
@@ -0,0 +1,99 @@
+    private BigDecimal price;
+    private Duration duration;
+    private String creationDate;
+    private String lastUpdateDate;
```
Let's store these date fields using Date class.

a1exanddrovich replied:
Now it has ZonedDateTime class to store date

NatalliaKhadunai commented on src/main/java/com/epam/esm/entity/Identifiable.java:
```diff
@@ -0,0 +1,4 @@
+package com.epam.esm.entity;
+
+public interface Identifiable {
```
Why do we need this interface?

a1exanddrovich replied:
To mark entities and for the Dao interface

NatalliaKhadunai commented on src/main/java/com/epam/esm/entity/SqlQueries.java:
```diff
@@ -0,0 +1,32 @@
+package com.epam.esm.entity;
```
This class should be located in some other package. It's not an entity

a1exanddrovich replied:
Created a package called sql and moved it there

NatalliaKhadunai commented on src/main/java/com/epam/esm/utils/QueryConstructor.java:
```diff
@@ -0,0 +1,77 @@
+        }
+
+        if (tagName != null) {
+            query.append(" HAVING gc.id IN (SELECT gc.id FROM gift_certificate gc JOIN gift_certificate_tag gct ON gct.gift_certificate_id = gc.id JOIN tag t ON gct.tag_id = t.id WHERE t.name = '").append(tagName).append("')");
```
Please move this query `(SELECT gc.id FROM gift_certificate ...` somewhere

a1exanddrovich replied:
Moved to the SqlQueries class

NatalliaKhadunai commented on src/main/java/com/epam/esm/service/GiftCertificateService.java:
```diff
@@ -0,0 +1,133 @@
+        return getGiftCertificate(lastInsertedId);
+    }
+
+    private void checkForTags(Set<Tag> newTags) {
```
As we've discussed we should return `Set<Tag>` here that is going to contain all the tags that were initalized. Can be changed to smth like this:
```
private Set<Tag> initializeTags(Set<String> incomingTags) {
        return incomingTags.stream()
            .map(tagName -> {
                Optional<Tag> existingTag = tagDao.findTagByName(tagName);
                if (existingTag.isPresent()) {
                    return existingTag.get();
                }
                else {
                    Tag newTag = new Tag(tagName);
                    long tagId = tagDao.create(newTag);
                    newTag.setId(tagId);
                    return newTag;
                }
            })
            .collect(Collectors.toSet());
    }
```

a1exanddrovich replied:
The method initializeTags() now returns a Set<Tag>

NatalliaKhadunai commented on src/main/java/com/epam/esm/service/GiftCertificateService.java:
```diff
@@ -0,0 +1,133 @@
+
+    private void connectCertificatesAndTags(long id, GiftCertificate giftCertificate) {
+        List<Long> tagIdsBeforeUpdate = giftCertificateTagDao.getIdsBeforeUpdate(id);
+        List<Long> tagIdsAfterUpdate = tagDao.getIdsAfterUpdate(giftCertificate.getTags());
```
I assume that by this time `giftCertificate.getTags()` should return you a fully initalized set of tags with tag id's. So why do we need to call this method?

a1exanddrovich replied:
After we implemented initializeTags() properly method we do not need the giftCertificate.getTags() at all

NatalliaKhadunai commented on src/main/java/com/epam/esm/exception/EntityAlreadyExistsException.java:
```diff
@@ -0,0 +1,7 @@
+package com.epam.esm.exception;
+
+public class EntityAlreadyExistsException extends Exception {
```
Why all these exceptions extend `Exception` class?

a1exanddrovich replied:
Since we are not going to catch the exceptions and handle them somehow in catch{} section, it should extend RuntimeExceprion class. This way I implemented it

NatalliaKhadunai commented on src/main/test/com/epam/esm/dao/GiftCertificateDaoTest.java:
```diff
@@ -0,0 +1,107 @@
+package com.epam.esm.dao;
```
By maven's convention tests should be located in a folder `root_folder\src\test\java\com\epam\...`.

a1exanddrovich replied:
Fixed

NatalliaKhadunai commented on src/main/test/com/epam/esm/service/GiftCertificateServiceTest.java:
```diff
@@ -0,0 +1,109 @@
+import static org.mockito.Matchers.anyLong;
+import static org.mockito.Mockito.*;
+
+public class GiftCertificateServiceTest {
```
We need more tests for the service layer. Per requirements test coverage for a service layer should be no less than 80 percent. Right now it's only 20%:
![image](https://user-images.githubusercontent.com/17936204/131496243-f8cc59c9-b660-4bb1-8cef-7f3cf00fe2c7.png)

a1exanddrovich replied:
Now it has 91% test coverage according to IntelliJ IDEA tools

NatalliaKhadunai commented on src/main/test/com/epam/esm/service/GiftCertificateServiceTest.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    @Test
+    public void testShouldCreateGiftCertificate() throws EntityNotExistsException, BadEntityException {
```
You should probably add more checks to some test methods. Like in this one and `testShouldUpdateGiftCertificate` you can compare some fields of the returned object and dto to see that entity objects were populated/created/updated correctly.

a1exanddrovich replied:
Added checks in test classes in order to ensure that test works properly

NatalliaKhadunai commented on src/main/test/com/epam/esm/dao/GiftCertificateDaoTest.java:
```diff
@@ -0,0 +1,107 @@
+        constructor = new QueryConstructor();
+        mapper = new GiftCertificateMapper();
+        giftCertificate = new GiftCertificate();
+        dao = new GiftCertificateDao(template, mapper, constructor);
```
You shouldn't create all these objects like this using `new` keyword. As we're using Spring we need to make use of it in test as well and create som test Spring context and inject beans.

NatalliaKhadunai commented on src/main/test/com/epam/esm/dao/GiftCertificateDaoTest.java:
```diff
@@ -0,0 +1,107 @@
+        dao.updateGiftCertificate(2, giftCertificate);
+
+        //then
+        verify(dao, times(1)).updateGiftCertificate(id, giftCertificate);
```
In this type of integration tests you shouldn't use `verify`, it's not a mock or spy to verify smth. You need to check the actual data in the DB.

a1exanddrovich replied:
verify removed

NatalliaKhadunai commented on src/main/test/com/epam/esm/dao/GiftCertificateDaoTest.java:
```diff
@@ -0,0 +1,107 @@
+        Optional<GiftCertificate> certificate = dao.findById(id);
+
+        //then
+        Assert.assertNotNull(certificate.get());
```
Please check some fields of the returned object.

a1exanddrovich replied:
Now it checks for correctness comparing fields, not just checking that returned entity is not Null

NatalliaKhadunai committed:
Natallia khadunai maven action (#3)

* Create maven action

* Delete .github/workflows directory

* Create maven.yml

NatalliaKhadunai committed:
Update maven.yml

Commit:
Test fixes performed

Commit:
Merge branch 'development' of github.com:a1exanddrovich/CertificateSystem into development

Commit:
Multi-project removed

Commit:
Final fixes

AliakseiYermachonak commented on pom.xml:
```diff
@@ -0,0 +1,156 @@
+      <version>${spring.version}</version>
+      <scope>test</scope>
+    </dependency>
+
```
empty lines

a1exanddrovich replied:
fixed

AliakseiYermachonak commented on pom.xml:
```diff
@@ -0,0 +1,156 @@
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>4.13.2</version>
```
is you use properties, why don't use them through all the file?

NatalliaKhadunai replied:
@a1exanddrovich please answer the comment.

a1exanddrovich replied:
now all the properties moved to the <properties/> section

AliakseiYermachonak commented on src/main/java/com/epam/esm/config/DevDataSourceConfig.java:
```diff
@@ -0,0 +1,20 @@
+    public DataSource dataSource() {
+        return new EmbeddedDatabaseBuilder().setType(EmbeddedDatabaseType.H2).addScript("test-data.sql").build();
+    }
+
```
code style

a1exanddrovich replied:
Fixed

AliakseiYermachonak commented on src/main/java/com/epam/esm/controller/BusinessExceptionHandler.java:
```diff
@@ -0,0 +1,51 @@
+    @ExceptionHandler(EntityNotExistsException.class)
+    public ResponseEntity<ExceptionResponse> handleEntityNotExistException(Locale locale) {
+        return new ResponseEntity<>(new ExceptionResponse(HttpStatus.NOT_FOUND.value(),
+                                                          messageSource.getMessage(NOT_EXIST_EXCEPTION_MESSAGE, new Object[]{}, locale)),
```
why do lines jump?

NatalliaKhadunai replied:
@a1exanddrovich please answer the comment.

a1exanddrovich replied:
I corrected that awful code style

AliakseiYermachonak commented on src/main/java/com/epam/esm/controller/GiftCertificatesController.java:
```diff
@@ -0,0 +1,111 @@
+     * @return {@link ResponseEntity} contained both {@link HttpStatus} status and {@link List} of {@link GiftCertificateDto}
+     */
+    @GetMapping(produces = JSON)
+    public ResponseEntity<List<GiftCertificateDto>> giftCertificates(@RequestParam(required = false) String tagName,
```
naming

a1exanddrovich replied:
The name fixed

Commit:
Fixed pom.xml and BusinessExceptionHandler

AliakseiYermachonak commented on src/main/java/com/epam/esm/dao/GiftCertificateTagDao.java:
```diff
@@ -0,0 +1,48 @@
+        template.update(SqlQueries.DISCONNECT_TAG_TO_CERTIFICATE, id, tagId);
+    }
+
+    public void createConnections(long giftCertificateId, Set<Tag> tags) {
```
why did you decide to have void here?

a1exanddrovich replied:
There is no need to return something. We just make tied certificates and relevant tags.

AliakseiYermachonak replied:
is here only one result action possible?

NatalliaKhadunai replied:
@a1exanddrovich, please don't leave comments unanswered! 
Should be fixed, but since aren't going to have this implementation in the next module, I'd say that we can leave it as is for now.

a1exanddrovich replied:
I assume we can get either a successful result or an unsuccessful one while tying given certificates to their tags

AliakseiYermachonak commented on src/main/java/com/epam/esm/service/GiftCertificateService.java:
```diff
@@ -0,0 +1,155 @@
+            }
+        }
+
+//        for (Tag tag : giftCertificate.getTags()) {
```
no commented code elements

NatalliaKhadunai replied:
@a1exanddrovich please fix the issue in the next module.

a1exanddrovich replied:
All the comments were removed

AliakseiYermachonak commented on src/main/java/com/epam/esm/utils/GiftCertificateDtoMapper.java:
```diff
@@ -0,0 +1,42 @@
+public class GiftCertificateDtoMapper {
+
+    public GiftCertificateDto map(GiftCertificate giftCertificate) {
+        GiftCertificateDto giftCertificateDto = new GiftCertificateDto(giftCertificate.getId(),
```
which pattern do we use in case a constructor has lot of parameters?

a1exanddrovich replied:
We use Builder pattern

NatalliaKhadunai replied:
@a1exanddrovich please note that we'll need to fix this in the next module.

a1exanddrovich replied:
We'll fix it

AliakseiYermachonak commented on pom.xml:
```diff
@@ -0,0 +1,156 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
is it a multi module project?

NatalliaKhadunai replied:
I've asked Aliaksei to implement the project using a single-module structure.

AliakseiYermachonak commented on src/test/java/com/epam/esm/config/TestConfig.java:
```diff
@@ -0,0 +1,25 @@
+package com.epam.esm.config;
```
not sure that test coverage s according to the task

a1exanddrovich replied:
Rest assured it has a 91% coverage according to IntelliJ IDEA tools

NatalliaKhadunai commented on src/main/java/com/epam/esm/config/SpringConfig.java:
```diff
@@ -0,0 +1,67 @@
+    @Bean
+    public LocaleChangeInterceptor localeChangeInterceptor() {
+        LocaleChangeInterceptor localeChangeInterceptor = new LocaleChangeInterceptor();
+        localeChangeInterceptor.setParamName(LOCALE);
```
Please use the `Accept-Language` header to set locale. Not URL parameter.

a1exanddrovich replied:
Fixe. Now we use "Accept-language" header instead

Commit:
Added determining locale via headers. Fixed some names

Commit:
Added entities e.g. User, Order, service and dao for them.
Migrated to JPA
Added pagination and some new conditions for GiftCertificate searching

Commit:
Finished 2 module

Commit:
Fixed some remarks

Commit:
Fixed some more remarks

Commit:
Fixed some more remarks

Commit:
Fixed pom.xml

Commit:
Fixed pom.xml

Commit:
added web.xml

Commit:
Removed web.xml
GiftCertificateDaoTest fixed e.g. it uses spring autowiring instead of new
